_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+`createIndexer` function also takes in Class references as input. This gives you more choice over the indexing. This is going to be a one-time thing, after this Hibernate search will keep entities in both sources in sync. Unless of course for some reason your database goes out of sync with Elasticsearch in which case indexing API might come in handy again. 
+
+With Elasticsearch integration you have two choices for writing queries:
+1. Hibernate search query DSL - It's a nice way to write Lucene queries. If you are familiar with Specification and 
+Criteria query API you will find it easy to get your head around it.
+2. Elasticsearch query - Hibernate search supports both Elasticsearch native queries and JSON queries.
+
+In this tutorial, we are only going to look at Hibernate Search query DSL. 
+
+Now let's say we want to write a query to fetch all records from `idx_post`  where either `body` or `hashtags` contain 
+word 'food'
+
+```java
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class SearchService {
+
+  private final EntityManager entityManager;
+  //.........
+  public List<Post> getPostBasedOnWord(String word){
+    FullTextEntityManager fullTextEntityManager = 
+              Search.getFullTextEntityManager(entityManager);
+
+    QueryBuilder qb = fullTextEntityManager
+              .getSearchFactory()
+              .buildQueryBuilder()
+              .forEntity(Post.class)
+              .get();
+
+    Query foodQuery = qb.keyword()
+              .onFields("body","hashTags")
+              .matching(word)
+              .createQuery();
+
+    FullTextQuery fullTextQuery = fullTextEntityManager
+                          .createFullTextQuery(foodQuery, Post.class);
+    return (List<Post>) fullTextQuery.getResultList();
+  }
+  //.........
+  
+}
+```
+
+Let me explain the above code line by line:
+1. First, we create an object of `FullTextEntitiyManager` which is a wrapper over our `EntityManager`
+2. Next, we create `QueryBuilder` for the index on which we want to perform a search. We also need to pass the entity class object in it.
+3. Using `QueryBuilder` we finally build our `Query`. 
+4. Next, we make use of the keyword query `keyword()` which allows us to look for a specific word in a field or fields. Finally, we pass the word that we want to search in the `matching` 
+function.
+4. Lastly, we wrap everything in `FullTextQuery` and fetch result list by calling `getResultList()` function.
+
+One thing to note here is that although we are performing a query on Elasticsearch, Hibernate will still fire a query on the database to fetch the full entity. Which make sense, because as we saw in the previous section we didn't store all the fields of 
+`Post` entity and those fields still needed to be retrieved. If you only wanted to fetch what's stored in your index anyway and think this database call is redundant, I know just the way. Keep reading on to find out.
+
+Let's move to the next query. 
```

Comment:
A sentence like this makes me think this could be its own sub section. Can you think of a title for that subsection? What makes this query different from the first?